functions/go/apply-setters/apply_setters.go

```diff
@@ -0,0 +1,241 @@
+		sv := setterValue(as.Setters, setterPattern)
+
+		// parse the setter value as yaml node
+		rn, err := yaml.Parse(sv)
+		if err != nil {
+			return err
+		}
+
+		// the setter value must parse as sequence node
+		if rn.YNode().Kind != yaml.SequenceNode {
+			return errors.Errorf("input to array setter must be an array of values, but found %q", sv)
+		}
+
+		node.Value.YNode().Content = rn.YNode().Content
+		node.Value.YNode().Style = yaml.FoldedStyle
+		return nil
+	})
+}
+
+// visitScalar accepts the input scalar node and performs following steps,
+// checks if the line comment of input scalar node has prefix SetterCommentIdentifier
+// resolves the setter values for the setter name in the comment
+// replaces the existing value of the scalar node with the new value
+func (as *ApplySetters) visitScalar(object *yaml.RNode) error {
+	if object == nil || object.YNode() == nil {
+		return nil
+	}
+	// perform a direct set of the field if it matches
+	setterPattern := extractSetterPattern(object)
+	if setterPattern == "" {
+		// the node is not tagged with setter pattern
+		return nil
+	}
+
+	curPattern := setterPattern
+	if !shouldSet(setterPattern, as.Setters) {
+		// this means there is no intent from user to modify this setter tagged resources
+		return nil
+	}
+
+	// replace the setter names in comment pattern with provided values
+	for _, setter := range as.Setters {
+		setterPattern = strings.ReplaceAll(
+			setterPattern,
+			fmt.Sprintf("${%s}", setter.Name),
+			fmt.Sprintf("%v", setter.Value),
+		)
+	}
+
+	// replace the remaining setter names in comment pattern with values derived from current
+	// field value, these values are not provided by user
+	currentSetterValues := currentSetterValues(curPattern, object.YNode().Value)
+	for setterName, setterValue := range currentSetterValues {
+		setterPattern = strings.ReplaceAll(
+			setterPattern,
+			fmt.Sprintf("${%s}", setterName),
+			fmt.Sprintf("%v", setterValue),
+		)
+	}
+
+	// check if there are unresolved setters and throw error
+	urs := unresolvedSetters(setterPattern)
+	if len(urs) > 0 {
+		return errors.Errorf("values for setters %v must be provided", urs)
+	}
+
+	object.YNode().Value = setterPattern
+	object.YNode().Tag = yaml.NodeTagEmpty
+	return nil
+}
+
+// shouldSet takes the setter pattern comment and setter values map and returns true
+// iff at least one of the setter names in the pattern match with the setter names
+// in input setterValues map
+func shouldSet(pattern string, setters []Setter) bool {
+	for _, s := range setters {
+		if strings.Contains(pattern, fmt.Sprintf("${%s}", s.Name)) {
+			return true
+		}
+	}
+	return false
+}
+
+// currentSetterValues takes pattern and value and returns setter names to values
+// derived using pattern matching
+// e.g. pattern = foo-${image}:${tag}-bar, value = foo-nginx:1.7.1-bar
+// returns {"image":"nginx", "tag":"1.7.1"}
+func currentSetterValues(pattern, value string) map[string]string {
```

Comment:
so if the user doesn't provide setter values in the function config, you use the current field value to resolve the setter? Why is that the desired UX?

Reply:
Here is the example for that https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/167/files#diff-393a72510bdc10e07b06ed9459efcb52cb52eeac5c5e006385ae64cb17d4302cR116

If there is a field 
```
image: nginx:1.7.1 # kpt-set: ${image}:${tag}
```

And if user wants to just update tag and not the image then the input would just be 
```
tag: 1.8.0
```

We need a way to identify the current value of ${image}. This method derives that.

Reply:
Where's this behavior documented? Have you considered the alternatives?

It's a form of per-reference defaulting.

Reply:
> Where's this behavior documented? Have you considered the alternatives?
> 
> It's a form of per-reference defaulting.

I have updated the setters doc with reasoning. I can iterate on future PRs if the decision is otherwise.